### PR TITLE
Backport spacing presets 

### PR DESCRIPTION
**Pull request by glendaviesnz**
~❗ BLOCKED by https://github.com/WordPress/wordpress-develop/pull/3204~ _Edit: #3204 has been merged._

Trac ticket: https://core.trac.wordpress.org/ticket/56467

### CODE REVIEW

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (andrewserong) [inline tests/phpunit/tests/theme/wpThemeJson.php:3503]: Based on feedback on the other PRs, for consistency with the other tests, I believe we need to add a `@ticket` line to these doc comments, so that it's linked to the trac ticket for the backports:

```
 * @ticket 56467
```

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
				trigger_error( __( 'Some of the theme.json settings.spacing.spacingScale values are invalid' ), E_USER_NOTICE );
```
`gutenberg` textdomains should be removed before backport.

🔢 Applies to other instances in the PR.

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
		$steps_mid_point = round( $spacing_scale['steps'] / 2, 0 );
```
While I'm a fan of parentheses, those pairs were redundant. 🙂

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php]: This a  little awkward to read. I suggest using `if/elseif/else` to break up the conditions.

```suggestion
			if ( '+' === $spacing_scale['operator'] ) {
				$current_step -= $spacing_scale['increment'];
			} elseif ( $spacing_scale['increment'] > 1 ) {
				$current_step /= $spacing_scale['increment'];
			} else {
				$current_step *= $spacing_scale['increment'];
			}
```

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
				/* translators: %s: Multiple of t-shirt sizing, eg. 2X-Small. */
```
- Needs a period at the end.
- Also, "t-shirt sizing" doesn't seem very useful for translators. Is there precedent for using this type of description in Core?

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php]: Core is removing single-character variables for readability. Can we replace `$x` with something else?

Related: `$x_small_count` should probably change as well. It's also somewhat ambiguous itself.

🔢 Applies to other instances in the PR.

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
			$slug -= 10;
```

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
			$slug += 10;
```

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJson.php:3500]: `@ticket` should precede `@dataProvider`, and the test description should be the first line.

[See here](https://core.trac.wordpress.org/ticket/56070) for the order of descriptions and annotations.

🔢 Applies to other instances in the PR.

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJson.php]:
```suggestion
	function test_should_set_spacing_sizes( $spacing_scale, $expected_output ) {
```

Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJson.php]: Test method arguments should be documented at the end of the test method's docblock.

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJson.php]: The empty lines between datasets can be removed.

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJson.php]:
```suggestion
				'spacing_scale'   => array(
```
🔢 Applies to other instances in the PR.

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJson.php]: Dataset names should be written with this in mind:

```
<test method name> with <dataset name>
```
In this instance (and without any suggestions in this review implemented), the output with the `--testdox` flag would be:
```
set spacing sizes with if_bottom_end_will_go_below_zero_should_add_extra_steps_above_medium_instead
```
So, while we're not writing an instruction manual on grammar, we _are_ telling a story of the expected behaviour. IMO, it's worth making this easier to read and understand what's being tested, and the expected outcome. However, if there is a documented practice in Core for writing dataset names this way, please let me know. 🙂

Also, dataset names don't need to have an `_` between words and can just be written as `the dataset name` for readability in source and CLI.

🔢 Applies to other instances in the PR.

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for handling this PR @glendaviesnz! I've left some thoughts below.

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (glendaviesnz) [reply tests/phpunit/tests/theme/wpThemeJson.php]: The method expects the key to be `spacingScale` as this is what comes back from the theme.json parsing

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php]: done

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php]: hmmmm, I think a linter/formatter somewhere went wild on that - fixed.

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php]: done

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php]: Have updated it to something hopefully more understandable across languages - we may be dropping the 2x-small style labels sometime soon anyway.

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php]: done

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php]: done

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php]: done

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (glendaviesnz) [reply tests/phpunit/tests/theme/wpThemeJson.php:3500]: done

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (glendaviesnz) [reply tests/phpunit/tests/theme/wpThemeJson.php]: done

Comment (glendaviesnz) [reply tests/phpunit/tests/theme/wpThemeJson.php]: done

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (glendaviesnz) [reply tests/phpunit/tests/theme/wpThemeJson.php]: done

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (glendaviesnz) [reply tests/phpunit/tests/theme/wpThemeJson.php]: Thanks for the explanation, I have reworded all of these in a way which hopefully makes more sense

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php:2943]: Since `$x` has been renamed to `$below_midpoint_count`, `$x_small_count` should be renamed to make it clear what it refers to. Does `$small_below_midpoint_count` make sense here, or?

Same for `$x_large_count`.

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJson.php]:
```suggestion
	public function test_set_spacing_sizes_should_detect_invalid_spacing_scale( $spacing_scale, $expected_output ) {
```

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJson.php]: With the above suggestion for renaming the test method, "spacing scale is " and "spacing scale has " can be removed from these datasets and will now read, for example:

```
<set spacing sizes should detect invalid spacing scale> with <missing operator value>
```

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (costdev) [reply tests/phpunit/tests/theme/wpThemeJson.php]: `spacingScale` in this context is the name of the test method's argument rather than the method being tested, which should match `$spacing_scale` in the test method's signature.

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the updates @glendaviesnz! A few more thoughts and [a comment](https://github.com/WordPress/wordpress-develop/pull/3255#discussion_r972659698) on the `spacingScale` discussion from my last review.

**Comment by glendaviesnz** [comment]
@costdev fyi there is now some discussion going  about whether the t-shirt size style labels will be dropped for 6.1, so need to wait for a decision on that and will then update this PR accordingly.

**Comment by costdev** [comment]
Thanks for the update @glendaviesnz!

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (glendaviesnz) [reply tests/phpunit/tests/theme/wpThemeJson.php]: 👍 That tidies it up nicely combined with the dataset renames.

File: tests/phpunit/tests/theme/wpThemeJson.php
Comment (glendaviesnz) [reply tests/phpunit/tests/theme/wpThemeJson.php]: done.

**Comment by glendaviesnz** [comment]
> spacingScale in this context is the name of the test method's argument rather than the method being tested, which should match $spacing_scale in the test method's signature.

🤦 so it is, doh! fixed.

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php:2943]: in this case the `x` relates to the number of `X`s in the X-Small count, eg. `X-Small, 2X-Small` so it sort of is meaningful rather than just a standard `$x` count variable .... still waiting on a decision if we retain this or not, but can rename it if we do retain it and you think that is still confusing

File: src/wp-includes/class-wp-theme-json.php
Comment (costdev) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
				'name' => $below_midpoint_count === $steps_mid_point - 1 ? __( 'Small' ) : sprintf( __( '%sX-Small' ), (string) $x_small_count ),
```
I believe Core uses type casts rather than `strval()/intval()/etc`. [See this ticket](https://core.trac.wordpress.org/ticket/42918) as an example.

🔢 Applies elsewhere in the PR.

**Comment by glendaviesnz** [comment]
I think this should be good to merge, pending @costdev double checking the switch to `(string)` casting

File: src/wp-includes/class-wp-theme-json.php
Comment (glendaviesnz) [reply src/wp-includes/class-wp-theme-json.php:2943]: We are  retaining the t-shirt style sizing for spacing scales with steps > 7 for now, but with the `$x` vars gone now, and the var used in this context `sprintf( __( '%sX-Small' ), (string) $x_small_count ),` I think the naming is clear enough to avoid confusion around it ... but let me know if you still think it should be renamed.

**Review by costdev — APPROVED** [review]
LGTM :+1: Thanks @glendaviesnz

File: src/wp-includes/class-wp-theme-json.php
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-theme-json.php]: Return should be `null` also.
```suggestion
	 * @return null|void
```

**Review by mukeshpanchal27 — COMMENTED** [review]
LGTM. left minor document change.

**Comment by audrasjb** [comment]
Committed in https://core.trac.wordpress.org/changeset/54272

**Pull request closed without merging**
